components/core/tests/test-ir_serializer.cpp

```diff
@@ -0,0 +1,147 @@
+#include <chrono>
+#include <filesystem>
+#include <string>
+#include <vector>
+
+#include <Catch2/single_include/catch2/catch.hpp>
+
+#include "../src/clp/ErrorCode.hpp"
+#include "../src/clp/ffi/ir_stream/decoding_methods.hpp"
+#include "../src/clp/FileReader.hpp"
+#include "../src/clp/ir/constants.hpp"
+#include "../src/clp/ir/LogEventDeserializer.hpp"
+#include "../src/clp/ir/LogEventSerializer.hpp"
+#include "../src/clp/streaming_compression/zstd/Decompressor.hpp"
+#include "../src/clp/Utils.hpp"
+
+using clp::ir::eight_byte_encoded_variable_t;
+using clp::ir::four_byte_encoded_variable_t;
+using std::string;
+using std::vector;
+
+using clp::ffi::ir_stream::deserialize_log_event;
+
+using std::chrono::milliseconds;
+using std::chrono::system_clock;
+using std::is_same_v;
+
+using clp::ErrorCode_Success;
+using clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+using clp::FileReader;
+using clp::ir::cIrFileExtension;
+using clp::ir::epoch_time_ms_t;
+using clp::ir::LogEventDeserializer;
+using clp::ir::LogEventSerializer;
+using clp::streaming_compression::zstd::Decompressor;
+
+namespace {
+template <typename encoded_variable_t>
+auto match_encoding_type(bool is_four_bytes_encoding) -> bool {
+    static_assert(
+            (is_same_v<encoded_variable_t, eight_byte_encoded_variable_t>)
+            || (is_same_v<encoded_variable_t, four_byte_encoded_variable_t>)
+    );
+
+    if constexpr (is_same_v<encoded_variable_t, eight_byte_encoded_variable_t>) {
+        return false == is_four_bytes_encoding;
+    } else {
+        return is_four_bytes_encoding;
+    }
+}
+}  // namespace
+
+/*
+ * The test case only covers four byte encoding because decompressor
+ * does not support eight bytes encoding yet.
+ */
+TEMPLATE_TEST_CASE(
+        "Serialize log events",
+        "[ir][serialize-log-event]",
+        four_byte_encoded_variable_t
+) {
+    LogEventSerializer<TestType> serializer;
+
+    // Test encoding with serializer
+    vector<string> var_strs
+            = {"4938",
+               std::to_string(INT32_MAX),
+               std::to_string(INT64_MAX),
+               "0.1",
+               "-25.519686",
+               "-25.5196868642755",
+               "-00.00",
+               "bin/python2.7.3",
+               "abc123"};
```

Comment:
is it necessary to use a vector? Can we replace `var_strs[var_idx++]` with the actual variable string?

Reply:
we can, but if I do "text" + "text". C++ will complain that it can't add two char[].
If I do "text" + std::string(4938). I get a clang-tidy warning of magic number, but defining separate constant and use it once is not any better than having a vector. any suggestion?